olive/passes/onnx/gptq_utils.py

```diff
@@ -0,0 +1,299 @@
+        )
+
+        general_pack_on_row(qweight_gpu, intweight_gpu, self.bits)
+        self.qweight = qweight_gpu.cpu()
+
+        self.pack_qzeros_odd(intzeros, device)
+
+        if self.orig_fp_weight is not None:
+            fw, _, _ = self.unpack()
+            assert (fw == self.orig_fp_weight.to(device)).all()
+
+    def pack_qzeros_even(self, intzeros, device):
+        intzeros = intzeros.int()
+        # why -1?
+        # zeros_cuda = (zeros - 1).to(device).int()
+        compatible_with_autogptq = int(os.environ.get("COMPATIBLE_WITH_AUTOGPTQ", "0"))
+        zeros_cuda = intzeros - compatible_with_autogptq
+        max_num_in_bits = 2**self.bits - 1
+        zeros_cuda = (zeros_cuda.byte() & max_num_in_bits).int()
+        qzeros_cuda = torch.zeros(
+            (intzeros.shape[0], intzeros.shape[1] // 32 * self.bits), dtype=torch.int32, device=device
+        )
+        qzeros_cuda = qzeros_cuda.T.contiguous()
+        zeros_cuda = zeros_cuda.T.contiguous()
+        pack_on_row_fast_248bit(qzeros_cuda, zeros_cuda, self.bits)
+        self.qzeros = qzeros_cuda.T.contiguous().cpu()
+
+    def pack_on_device_for_even_bits(self, intweight_gpu, intzeros_T):
+        self.act_order = self.g_idx[: self.groupsize // self.bits].sum().item() != 0
+        assert self.act_order is False, "please set the pack_model to others like [GPTQ, AUTO] and try again."
+        intzeros_pt = intzeros_T.T.byte()
+        scales_pt = self.scales.T.to(intweight_gpu.device)
+        intweight_pt = intweight_gpu.byte()
+        block_size = self.groupsize
+
+        rows, cols = intweight_pt.shape
+        blob_size = block_size // 2
+        k_blocks = (rows + block_size - 1) // block_size
+        padded_rows = k_blocks * block_size
+        pad_len = padded_rows - rows
+        if pad_len > 0:
+            intweight_pt = torch.nn.functional.pad(intweight_pt, (0, 0, 0, pad_len), "constant", 0)
+            intzeros_pt = torch.nn.functional.pad(intzeros_pt, (0, 0, 0, pad_len), "constant", 0)
+
+        intzeros_pt = (intzeros_pt[:, 0::2]) | (intzeros_pt[:, 1::2] << 4)
+        intzeros_pt = intzeros_pt.reshape(-1)
+
+        intweight_pt_t = intweight_gpu.T
+        intweight_pt_t = (intweight_pt_t[:, 0::2]) | (intweight_pt_t[:, 1::2] << 4)
+        intweight_pt_t = intweight_pt_t.reshape(cols, k_blocks, blob_size)
+
+        scales_pt = scales_pt.reshape(-1)
+
+        assert self.qweight.shape == intweight_pt_t.shape
+        assert self.qzeros.shape == intzeros_pt.shape
+
+        self.scales = scales_pt.contiguous()
+        self.qweight = intweight_pt_t.contiguous().byte()
+        self.qzeros = intzeros_pt.contiguous().byte()
+
+    def accelerate_pack_on_device(self, linear, scales, zeros, g_idx=None, device="cuda"):
+        scales = scales.to(device)
+        zeros = zeros.to(device)
+        layer_weight = linear.weight.data.to(device)
```

Comment:
as we discussed, make these logic device agnostic by depends on the input tensor devices instead of explicitly to the device in the method.